sklearn/neighbors/lmnn.py

```diff
@@ -0,0 +1,1345 @@
+        ind = np.unique(np.concatenate((ind_a, ind_b)))
+
+        if len(ind):
+            ind_plus_offset = ind + chunk.start * X_b.shape[0]
+            imp_indices.extend(ind_plus_offset)
+
+            if return_distance:
+                # We only need to do clipping if we return the distances.
+                # This adds another ~8% time of computation
+                distances_chunk = distances_ab.ravel()[ind]
+                # Clip only the indexed (unique) distances
+                np.maximum(distances_chunk, 0, out=distances_chunk)
+                imp_distances.extend(distances_chunk)
+
+    imp_indices = np.asarray(imp_indices)
+
+    if return_distance:
+        return imp_indices, np.asarray(imp_distances)
+    else:
+        return imp_indices
+
+
+def _compute_push_loss(X, target_neighbors, dist_tn, impostors_graph):
+    """
+
+    Parameters
+    ----------
+    X : array, shape (n_samples, n_features)
+        The training input samples.
+
+    target_neighbors : array, shape (n_samples, n_neighbors)
+        Indices of target neighbors of each sample.
+
+    dist_tn : array, shape (n_samples, n_neighbors)
+        (Squared) distances of samples to their target neighbors.
+
+    impostors_graph : coo_matrix, shape (n_samples, n_samples)
+        Element (i, j) is the distance between sample i and j if j is an
+        impostor to i, otherwise zero.
+
+    Returns
+    -------
+    loss : float
+        The push loss caused by the given target neighbors and impostors.
+
+    grad : array, shape (n_features, n_features)
+        The gradient of the push loss.
+
+    n_active_triplets : int
+        The number of active triplet constraints.
+
+    """
+
+    n_samples, n_neighbors = dist_tn.shape
+    imp_row = impostors_graph.row
+    imp_col = impostors_graph.col
+    dist_impostors = impostors_graph.data
+
+    loss = 0
+    shape = (n_samples, n_samples)
+    A0 = csr_matrix(shape)
+    sample_range = range(n_samples)
+    n_active_triplets = 0
+    for k in range(n_neighbors - 1, -1, -1):
+        loss1 = np.maximum(dist_tn[imp_row, k] - dist_impostors, 0)
+        ac, = np.where(loss1 > 0)
+        n_active_triplets += len(ac)
+        A1 = csr_matrix((2 * loss1[ac], (imp_row[ac], imp_col[ac])), shape)
+
+        loss2 = np.maximum(dist_tn[imp_col, k] - dist_impostors, 0)
+        ac, = np.where(loss2 > 0)
+        n_active_triplets += len(ac)
+        A2 = csc_matrix((2 * loss2[ac], (imp_row[ac], imp_col[ac])), shape)
+
+        val = (A1.sum(1).ravel() + A2.sum(0)).getA1()
+        A3 = csr_matrix((val, (sample_range, target_neighbors[:, k])), shape)
+        A0 = A0 - A1 - A2 + A3
+        loss += np.dot(loss1, loss1) + np.dot(loss2, loss2)
+
+    grad = _sum_weighted_outer_differences(X, A0)
+
+    return loss, grad, n_active_triplets
+
+
+##########################
+# Some helper functions #
+#########################
+
+def _euclidean_distances_without_checks(X, Y=None, Y_norm_squared=None,
```

Comment:
Can you just use `sklearn.config_context(assume_finite=True)` to avoid reimplementing this?

Reply:
If I remember correctly, you said this is not thread-safe. Also the original `euclidean_distances` function does not have the `clip` option which if `True` (default) adds ~8% to the total runtime.

Reply:
Okay, I suppose. Might add this to utilise.extmath, then?

Reply:
Sure, should I open another PR for that?

Reply:
No, we would usually only add a util when it was being used in some context anyway; but avoiding duplication between here and metrics.pairwise, as far as possible, would be a good idea.